{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "recovered-ensemble",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # metadata - normal illumina adrenal 12k\n",
    "# d = '/share/crsp/lab/seyedam/share/Heidi_Liz/adrenal/fastq/deep/'\n",
    "# depth = '12k'\n",
    "# batch = 'normal'\n",
    "# assay = 'illumina'\n",
    "# tissue = 'adrenal'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "hindu-leone",
   "metadata": {},
   "outputs": [],
   "source": [
    "# metadata - 8mo illumina adrenal 12k\n",
    "d = '/share/crsp/lab/seyedam/share/Heidi_Liz/bl6_5x_cast_adr_gas_heart/fastq/shallow/'\n",
    "batch = '8mo'\n",
    "assay = 'illumina'\n",
    "tissue = 'adrenal'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "threaded-beatles",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "id": "responsible-credits",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_metadata(f,\n",
    "                 batch='normal'):\n",
    "    \"\"\"\n",
    "    Return tissue, age, sex, replicate, and genotype metadata\n",
    "    from a filename.\n",
    "    \n",
    "    Parameters:\n",
    "        f (str): Base filename\n",
    "        eight_mo (str): Whether this is 8mo or normal exp.\n",
    "    \"\"\"\n",
    "    if batch != '8mo':\n",
    "        # R2s have one extra set of characters\n",
    "        if 'R2' in f:\n",
    "            tissue, age, sex, rep, _ = f.split('_')\n",
    "            rep = rep.split('.')[0]\n",
    "        else:\n",
    "            tissue, age, sex, rep = f.split('_')\n",
    "            rep = rep.split('.')[0]\n",
    "        geno = 'B'\n",
    "    else:\n",
    "        age = '8mo'\n",
    "        # R2s have one extra set of characters\n",
    "        if 'R2' in f:\n",
    "            geno, tissue, rep, _ = f.split('_')\n",
    "            rep = rep.split('.')[0]\n",
    "        else:\n",
    "            geno, tissue, rep = f.split('_')\n",
    "            rep = rep.split('.')[0]\n",
    "        sex = np.nan\n",
    "    return tissue, age, sex, rep, geno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "id": "above-reminder",
   "metadata": {},
   "outputs": [],
   "source": [
    "# files = pd.read_csv('test_fastqs.txt', header=None)\n",
    "# files.columns = ['fname']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "id": "assumed-combination",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_ref_data():\n",
    "    refs = dict()\n",
    "    \n",
    "    refs['age'] = pd.read_csv('refs/age_metadata.tsv', sep='\\t')\n",
    "    \n",
    "    refs['sex'] = pd.read_csv('refs/sex_metadata.tsv', sep='\\t')\n",
    "    \n",
    "    refs['tissue'] = pd.read_csv('refs/tissue_metadata.tsv', sep='\\t')\n",
    "    refs['tissue'].drop('donor', axis=1, inplace=True)\n",
    "    \n",
    "    refs['rep'] = pd.read_csv('refs/rep_metadata.tsv', sep='\\t')\n",
    "    refs['rep'].short = refs['rep'].short.astype(str)\n",
    "    \n",
    "    refs['geno'] = pd.read_csv('refs/genotype_metadata.tsv', sep='\\t')\n",
    "    \n",
    "    refs['assay'] = pd.read_csv('refs/assay_metadata.tsv', sep='\\t')\n",
    "\n",
    "    # limit library metadata based on what we're looking for?\n",
    "    refs['lib'] = pd.read_csv('refs/lib_metadata.tsv', sep='\\t')\n",
    "    cols = ['Sample ID', 'Tissue', 'depth2', 'batch', 'run_number', 'Date shipped',\n",
    "            'Fragment size (bp)', '# of input cells', 'Sex', 'Path to R1 fastq']\n",
    "    refs['lib'] = refs['lib'][cols]\n",
    "    refs['lib']['Path to R1 fastq'] = refs['lib']['Path to R1 fastq'].astype(str)\n",
    "    refs['lib']['dir'] = refs['lib'].apply(lambda x: os.path.dirname(x['Path to R1 fastq']), axis=1)\n",
    "    refs['lib'].rename({'depth2':'depth',\n",
    "                        'Date shipped': 'date_obtained', \n",
    "                        'Fragment size (bp)': 'average_fragment_size',\n",
    "                        '# of input cells': 'nucleic_acid_starting_quantity'}, axis=1, inplace=True)\n",
    "#     refs['lib'].date_obtained = refs['lib'].date_obtained.dt.strftime('%Y-%m-%d')\n",
    "    refs['lib'].date_obtained = refs['lib'].date_obtained.astype('datetime64')\n",
    "    \n",
    "    return refs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "id": "figured-market",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = refs['lib'][['batch', 'depth', 'dir', 'Sample ID']].groupby(['batch', 'depth', 'dir']).count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "id": "therapeutic-sherman",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>depth</th>\n",
       "      <th>dir</th>\n",
       "      <th>Sample ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8mo</td>\n",
       "      <td>12k</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/Heidi_Liz/bl6_5x...</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8mo</td>\n",
       "      <td>12k</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/Heidi_Liz/bl6_5x...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8mo</td>\n",
       "      <td>2k</td>\n",
       "      <td></td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8mo</td>\n",
       "      <td>2k</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/Heidi_Liz/bl6_5x...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>normal</td>\n",
       "      <td>12k</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/Heidi_Liz/adrena...</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    batch depth                                                dir  Sample ID\n",
       "0     8mo   12k  /share/crsp/lab/seyedam/share/Heidi_Liz/bl6_5x...         24\n",
       "1     8mo   12k  /share/crsp/lab/seyedam/share/Heidi_Liz/bl6_5x...         16\n",
       "2     8mo    2k                                                            24\n",
       "3     8mo    2k  /share/crsp/lab/seyedam/share/Heidi_Liz/bl6_5x...         16\n",
       "4  normal   12k  /share/crsp/lab/seyedam/share/Heidi_Liz/adrena...         28"
      ]
     },
     "execution_count": 298,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for ind, entry in temp:\n",
    "    if entry.dir == '':\n",
    "        continue\n",
    "    else:\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "id": "seventh-somewhere",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # find all the fastqs - test\n",
    "# ext = r'{}*.fastq.gz'.format(d)\n",
    "\n",
    "# ofile = 'test_fastqs.txt'\n",
    "# ofile = open(ofile, 'w')\n",
    "# for f in glob.glob(ext):\n",
    "#     ofile.write(f+'\\n')\n",
    "    \n",
    "    \n",
    "    \n",
    "# ofile.close()    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "id": "reverse-attack",
   "metadata": {},
   "outputs": [],
   "source": [
    "refs = load_ref_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "id": "defined-transcription",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/share/crsp/lab/seyedam/share/Heidi_Liz/adrenal/fastq/shallow'"
      ]
     },
     "execution_count": 291,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refs['lib']['dir'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "elect-participant",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = files.copy(deep=True)\n",
    "\n",
    "# add library information\n",
    "df['depth'] = depth\n",
    "df['batch'] = batch\n",
    "df['assay'] = assay\n",
    "\n",
    "# add consistent info\n",
    "df['lab'] = 'ali-mortazavi'\n",
    "df['award'] = 'UM1HG009443'\n",
    "df['nucleic_acid_term_name'] = 'RNA'\n",
    "df['construction_method'] = 'Parse Single Cell Whole Transcriptome Kit'\n",
    "df['nucleic_acid_starting_quantity_units'] = 'cells'\n",
    "df['technical_replicate_number'] = 1\n",
    "\n",
    "df['short_fname'] = df.apply(lambda x: os.path.basename(x.fname), axis=1)\n",
    "df[['tissue', 'age', 'sex', 'rep', 'geno']] = df.apply(lambda x: get_metadata(x.short_fname, batch=batch), axis=1, result_type='expand')\n",
    "\n",
    "if batch == 'normal':\n",
    "    df['Sample ID'] = df.tissue+'_'+df.age+'_'+df.sex+'_'+df.rep\n",
    "    df['biological_replicate_number'] = df.rep\n",
    "elif batch == '8mo':\n",
    "    df['Sample ID'] = df.geno+'_'+df.tissue+'_'+df.rep\n",
    "\n",
    "# tissue-related metadata\n",
    "df = df.merge(refs['tissue'], how='left', left_on='tissue', right_on='short')\n",
    "\n",
    "# library-related metadata\n",
    "df['tissue_merge'] = df.tissue_desc.str.capitalize()\n",
    "df = df.merge(refs['lib'], how='left', on=['Sample ID', 'depth', 'batch'])\n",
    "\n",
    "# 8mo data needs to pull sex from the spreadsheet\n",
    "if batch == '8mo':\n",
    "    df['sex'] = df['Sex']\n",
    "\n",
    "# age-related metadata\n",
    "df = df.merge(refs['age'], how='left', left_on='age', right_on='short')\n",
    "\n",
    "# sex-related metadata\n",
    "df = df.merge(refs['sex'], how='left', left_on='sex', right_on='short')\n",
    "\n",
    "# replicate-related metadata\n",
    "df = df.merge(refs['rep'], how='left', left_on='rep', right_on='short')\n",
    "\n",
    "# 8mo data needs to pull rep from metadata\n",
    "if batch == '8mo':\n",
    "    df['biological_replicate_number'] = df.replicate\n",
    "    df['rep'] = df['replicate'].astype('str')\n",
    "\n",
    "# genotype-related metadata\n",
    "df = df.merge(refs['geno'], how='left', left_on='geno', right_on='short')\n",
    "\n",
    "# assay-related metadata\n",
    "df = df.merge(refs['assay'], how='left', left_on='assay', right_on='short')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "id": "integral-cooling",
   "metadata": {},
   "outputs": [],
   "source": [
    "# biosample\n",
    "b = df.copy(deep=True)\n",
    "\n",
    "# create the alias\n",
    "b['aliases'] = 'ali-mortazavi:biosamp_'+b.alias_genotype+ \\\n",
    "               '_'+b.tissue_desc+ \\\n",
    "               '_'+b.age_desc+ \\\n",
    "               '_'+b.model_organism_sex+ \\\n",
    "               '_'+b.rep\n",
    "\n",
    "# create the description\n",
    "b[['desc_genotype', 'tissue_desc', 'age_desc', 'model_organism_sex', 'rep_desc']]\n",
    "b['description'] = b.desc_genotype+ \\\n",
    "                   ' '+b.tissue_desc+ \\\n",
    "                   ' '+b.age_desc+ \\\n",
    "                   ' '+b.model_organism_sex+ \\\n",
    "                   ' '+b.rep\n",
    "\n",
    "cols = ['biosample_ontology', 'organism', 'subcellular_fraction_term_name',\n",
    "        'donor', 'source', 'model_organism_sex', 'model_organism_age',\n",
    "        'model_organism_age_units', 'date_obtained', 'description',\n",
    "        'aliases', 'lab', 'award', 'Sample ID']\n",
    "b = b[cols]\n",
    "b.drop_duplicates(inplace=True)\n",
    "\n",
    "# merge in alias with orig. df to have access for libraries\n",
    "b_alias = b[['aliases', 'Sample ID']]\n",
    "b_alias.rename({'aliases': 'biosample'}, axis=1, inplace=True)\n",
    "df = df.merge(b_alias, how='left', on='Sample ID')\n",
    "b.drop('Sample ID', axis=1, inplace=True)\n",
    "b.drop_duplicates(inplace=True)\n",
    "\n",
    "fname = '{}_{}_biosample.tsv'.format(tissue, batch)\n",
    "b.to_csv(fname, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "fitting-norwegian",
   "metadata": {},
   "outputs": [],
   "source": [
    "# experiment\n",
    "e = df.copy(deep=True)\n",
    "\n",
    "# create the alias\n",
    "e['aliases'] = 'ali-mortazavi:experiment_'+e.alias_assay+ \\\n",
    "               '_'+e.alias_genotype+ \\\n",
    "               '_'+e.tissue_desc+ \\\n",
    "               '_'+e.age_desc+ \\\n",
    "               '_'+e.model_organism_sex+ \\\n",
    "               '_'+e.batch+ \\\n",
    "               '_'+e.depth\n",
    "\n",
    "# create the description\n",
    "e['description'] = e.desc_assay+ \\\n",
    "                   ' '+e.desc_genotype+ \\\n",
    "                   ' '+e.tissue_desc+ \\\n",
    "                   ' '+e.age_desc+ \\\n",
    "                   ' '+e.model_organism_sex+ \\\n",
    "                   ' '+e.batch+ \\\n",
    "                   ' '+e.depth\n",
    "\n",
    "cols = ['aliases', 'biosample_ontology', 'description', \\\n",
    "        'assay_term_name', 'lab', 'award', 'Sample ID']\n",
    "e = e[cols]\n",
    "e.drop_duplicates(inplace=True)\n",
    "\n",
    "# merge in alias with orig. df to have access for libraries\n",
    "e_alias = e[['aliases', 'Sample ID']]\n",
    "e_alias.rename({'aliases': 'experiment'}, axis=1, inplace=True)\n",
    "df = df.merge(e_alias, how='left', on='Sample ID')\n",
    "e.drop('Sample ID', axis=1, inplace=True)\n",
    "e.drop_duplicates(inplace=True)\n",
    "\n",
    "fname = '{}_{}_{}_{}_experiment.tsv'.format(tissue, batch, assay, depth)\n",
    "e.to_csv(fname, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "noticed-omega",
   "metadata": {},
   "outputs": [],
   "source": [
    "# library\n",
    "l = df.copy(deep=True)\n",
    "\n",
    "# create the alias\n",
    "l['aliases'] = 'ali-mortazavi:library_'+l.alias_assay+ \\\n",
    "               '_'+l.alias_genotype+ \\\n",
    "               '_'+l.tissue_desc+ \\\n",
    "               '_'+l.age_desc+ \\\n",
    "               '_'+l.model_organism_sex+ \\\n",
    "               '_'+l.depth+ \\\n",
    "               '_'+l.batch+ \\\n",
    "               '_'+l.rep\n",
    "\n",
    "# # create the description\n",
    "# l['description'] = l.desc_assay+ \\\n",
    "#                    ' '+l.desc_genotype+ \\\n",
    "#                    ' '+l.tissue_desc+ \\\n",
    "#                    ' '+l.age_desc+ \\\n",
    "#                    ' '+l.model_organism_sex+ \\\n",
    "#                    ' '+l.depth+ \\\n",
    "#                    ' '+l.rep\n",
    "\n",
    "cols = ['aliases', 'biosample', 'nucleic_acid_term_name', \n",
    "        'documents', 'construction_method', 'nucleic_acid_starting_quantity',\n",
    "        'nucleic_acid_starting_quantity_units', 'strand_specificity',\n",
    "        'average_fragment_size', 'lab', 'award', 'Sample ID']\n",
    "l = l[cols]\n",
    "l.drop_duplicates(inplace=True)\n",
    "\n",
    "# merge in alias with orig. df to have access for replicates, files\n",
    "l_alias = l[['aliases', 'Sample ID']]\n",
    "l_alias.rename({'aliases': 'library'}, axis=1, inplace=True)\n",
    "df = df.merge(l_alias, how='left', on='Sample ID')\n",
    "l.drop('Sample ID', axis=1, inplace=True)\n",
    "l.drop_duplicates(inplace=True)\n",
    "\n",
    "fname = '{}_{}_{}_{}_library.tsv'.format(tissue, batch, assay, depth)\n",
    "l.to_csv(fname, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "id": "private-bacon",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replicate \n",
    "r = df.copy(deep=True)\n",
    "\n",
    "# create the alias\n",
    "r['aliases'] = 'ali-mortazavi:replicate_'+r.alias_assay+ \\\n",
    "               '_'+r.alias_genotype+ \\\n",
    "               '_'+r.tissue_desc+ \\\n",
    "               '_'+r.age_desc+ \\\n",
    "               '_'+r.model_organism_sex+ \\\n",
    "               '_'+r.depth+ \\\n",
    "               '_'+r.batch+ \\\n",
    "               '_'+r.rep\n",
    "\n",
    "cols = ['aliases', 'library', 'experiment',\n",
    "        'biological_replicate_number', 'technical_replicate_number', \n",
    "        'Sample ID']\n",
    "r = r[cols]\n",
    "r.drop_duplicates(inplace=True)\n",
    "\n",
    "# merge in alias with orig. df to have access for files\n",
    "r_alias = r[['aliases', 'Sample ID']]\n",
    "r_alias.rename({'aliases': 'replicate'}, axis=1, inplace=True)\n",
    "df.drop('replicate', axis=1, inplace=True)\n",
    "df = df.merge(r_alias, how='left', on='Sample ID')\n",
    "r.drop('Sample ID', axis=1, inplace=True)\n",
    "r.drop_duplicates(inplace=True)\n",
    "\n",
    "fname = '{}_{}_{}_{}_replicate.tsv'.format(tissue, batch, assay, depth)\n",
    "r.to_csv(fname, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "id": "detected-clothing",
   "metadata": {},
   "outputs": [],
   "source": [
    "if assay == 'illumina':\n",
    "    \n",
    "    # r1 fastqs\n",
    "    r1 = df.copy(deep=True)\n",
    "\n",
    "    # create the alias\n",
    "    r1['aliases'] = 'ali-mortazavi:r1_fastq_'+r1.alias_assay+ \\\n",
    "                   '_'+r1.alias_genotype+ \\\n",
    "                   '_'+r1.tissue_desc+ \\\n",
    "                   '_'+r1.age_desc+ \\\n",
    "                   '_'+r1.model_organism_sex+ \\\n",
    "                   '_'+r1.depth+ \\\n",
    "                   '_'+r1.batch+ \\\n",
    "                   '_'+r1.rep\n",
    "\n",
    "    # rename experiment and fname\n",
    "    r1.rename({'experiment': 'dataset',\n",
    "               'fname': 'submitted_file_name'}, axis=1, inplace=True)\n",
    "\n",
    "    # r1-specific things\n",
    "    r1['file_format'] = 'fastq'\n",
    "    r1['output_type'] = 'reads'\n",
    "    r1['read_length'] = 115\n",
    "    r1['run_type'] = 'single-ended'\n",
    "    \n",
    "    # limit only to r1    \n",
    "    r1 = r1.loc[~r1.submitted_file_name.str.contains('_R2.fastq.gz')]\n",
    "    \n",
    "    cols = ['aliases', 'submitted_file_name', 'dataset',\n",
    "            'replicate', 'file_format', 'output_type',\n",
    "            'platform', 'lab', 'award', 'read_length',\n",
    "            'run_type', 'Sample ID']\n",
    "    r1 = r1[cols]\n",
    "    r1.drop_duplicates(inplace=True)\n",
    "\n",
    "    # merge in alias with orig. df to have access for r1s\n",
    "    r1_alias = r1[['aliases', 'Sample ID']]\n",
    "    r1_alias.rename({'aliases': 'index_of'}, axis=1, inplace=True)\n",
    "    df = df.merge(r1_alias, how='left', on='Sample ID')\n",
    "    r1.drop('Sample ID', axis=1, inplace=True)\n",
    "    r1.drop_duplicates(inplace=True)\n",
    "    \n",
    "    fname = '{}_{}_{}_{}_fastq_r1.tsv'.format(tissue, batch, assay, depth)\n",
    "    r1.to_csv(fname, sep='\\t', index=False)\n",
    "    \n",
    "    # r2 fastqs\n",
    "    r2 = df.copy(deep=True)\n",
    "    \n",
    "    # create the alias\n",
    "    r2['aliases'] = 'ali-mortazavi:r2_fastq_'+r2.alias_assay+ \\\n",
    "                   '_'+r2.alias_genotype+ \\\n",
    "                   '_'+r2.tissue_desc+ \\\n",
    "                   '_'+r2.age_desc+ \\\n",
    "                   '_'+r2.model_organism_sex+ \\\n",
    "                   '_'+r2.depth+ \\\n",
    "                   '_'+r2.batch+ \\\n",
    "                   '_'+r2.rep\n",
    "    \n",
    "    # rename experiment and fname\n",
    "    r2.rename({'experiment': 'dataset',\n",
    "               'fname': 'submitted_file_name'}, axis=1, inplace=True)\n",
    "    \n",
    "    # r2-specific things\n",
    "    r2['file_format'] = 'fastq'\n",
    "    r2['output_type'] = 'index reads'\n",
    "    r2['read_length'] = 86\n",
    "    r2['run_type'] = 'single-ended'\n",
    "    r2['read_structure'] = '{\"sequence_element\": \"UMI\", \"start\": 1, \"end\": 10},{\"sequence_element\": \"cell barcode\", \"start\": 11, \"end\": 18},{\"sequence_element\": \"cell barcode\", \"start\": 49, \"end\": 56},{\"sequence_element\": \"cell barcode\", \"start\": 79, \"end\": 86}'\n",
    "    \n",
    "    # limit only to r2\n",
    "    r2 = r2.loc[r2.submitted_file_name.str.contains('_R2.fastq.gz')]\n",
    "    \n",
    "    cols = ['aliases', 'dataset', 'submitted_file_name', \n",
    "            'replicate', 'file_format', 'output_type', \n",
    "            'platform', 'lab', 'award', 'read_length', \n",
    "            'run_type', 'index_of', 'read_structure', 'Sample ID']\n",
    "    \n",
    "    # merge in alias with orig. df to have access for downstream files\n",
    "    r2_alias = r2[['aliases', 'Sample ID']]\n",
    "    r2_alias.rename({'aliases': 'r2_fastq'}, axis=1, inplace=True)\n",
    "    df = df.merge(r2_alias, how='left', on='Sample ID')\n",
    "    df.rename({'index_of': 'r1_fastq'}, axis=1, inplace=True)\n",
    "    r2.drop('Sample ID', axis=1, inplace=True)\n",
    "    r2.drop_duplicates(inplace=True)\n",
    "    \n",
    "    fname = '{}_{}_{}_{}_fastq_r2.tsv'.format(tissue, batch, assay, depth)\n",
    "    r2.to_csv(fname, sep='\\t', index=False)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "classical-redhead",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
