{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "valid-spouse",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from argparse import ArgumentParser\n",
    "import os\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "heavy-regression",
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_df = pd.read_csv('tissue_metadata.tsv', sep='\\t')\n",
    "age_df = pd.read_csv('age_metadata.tsv', sep='\\t')\n",
    "sex_df = pd.read_csv('sex_metadata.tsv', sep='\\t')\n",
    "rep_df = pd.read_csv('rep_metadata.tsv', sep='\\t')\n",
    "\n",
    "\n",
    "tissue_df.set_index('short', inplace=True)\n",
    "age_df.set_index('short', inplace=True)\n",
    "sex_df.set_index('short', inplace=True)\n",
    "rep_df.set_index('short', inplace=True)\n",
    "\n",
    "\n",
    "lab = 'ali-mortazavi'\n",
    "award = 'UM1HG009443'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "going-sword",
   "metadata": {},
   "outputs": [],
   "source": [
    "lib_meta = pd.read_csv('adrenal_metadata.tsv', sep='\\t')\n",
    "lib_meta['# predicted nuclei'] = lib_meta['# predicted nuclei'].astype('int')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "floppy-values",
   "metadata": {},
   "outputs": [],
   "source": [
    "long = False\n",
    "opref = 'test/adrenal'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "legitimate-horizon",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test/*.fastq.gz\n",
      "A_18m_M_2\n",
      "A_10_F_1\n",
      "A_10_F_2\n"
     ]
    }
   ],
   "source": [
    "d = 'test'+'/'\n",
    "ext = r'{}*.fastq.gz'.format(d)\n",
    "\n",
    "biosamp_sub = pd.DataFrame()\n",
    "exp_sub = pd.DataFrame() # only one for each pair of reps!\n",
    "lib_sub = pd.DataFrame()\n",
    "rep_sub = pd.DataFrame()\n",
    "file_sub = pd.DataFrame()\n",
    "\n",
    "\n",
    "print(ext)\n",
    "for f in glob.glob(ext):\n",
    "    tissue, age, sex, rep = get_metadata(f)\n",
    "    sample_id = get_sample_id(f)\n",
    "    print(sample_id)\n",
    "    \n",
    "    # assemble metadata for this sample\n",
    "    temp = tissue_df.loc[tissue]\n",
    "    temp = pd.concat([temp, sex_df.loc[sex]], axis=0).to_frame()\n",
    "    temp = pd.concat([temp, age_df.loc[age]], axis=0)\n",
    "    temp = pd.concat([temp, rep_df.loc[int(rep)]], axis=0)\n",
    "    temp = temp.transpose()\n",
    "    \n",
    "    # add biosample\n",
    "    biosample = temp.copy(deep=True)\n",
    "    desc = 'B6Cast F1 {} {} {} {}'.format(temp.age_desc.values[0], temp.sex.values[0], \n",
    "                                          temp.tissue_desc.values[0], temp.rep_desc.values[0])\n",
    "    biosamp_alias = 'ali-mortazavi:biosample_{}_{}_{}_{}'.format(temp.age_desc.values[0], temp.sex.values[0],\n",
    "                                                         temp.tissue_desc.values[0], temp.rep.values[0])\n",
    "    biosample['date_obtained'] = lib_meta.loc[lib_meta['Sample ID'] == sample_id, 'Date shipped'].values[0]\n",
    "    biosample['description'] = desc\n",
    "    biosample['aliases'] = biosamp_alias\n",
    "    biosample['lab'] = lab\n",
    "    biosample['award'] = award\n",
    "\n",
    "    biosamp_sub = pd.concat([biosamp_sub, biosample], axis=0)\n",
    "    \n",
    "    # add experiment\n",
    "    exp = temp.copy(deep=True)\n",
    "    if long == False:\n",
    "        desc = 'Short-read Split-seq B6Cast F1 {} {} {}'.format(temp.age_desc.values[0], temp.sex.values[0], \n",
    "                                              temp.tissue_desc.values[0])\n",
    "        exp_alias = 'ali-mortazavi:exp_sr_{}_{}_{}'.format(temp.age_desc.values[0], temp.sex.values[0],\n",
    "                                                     temp.tissue_desc.values[0])\n",
    "    else:\n",
    "        desc = 'Long-read Split-seq B6Cast F1 {} {} {}'.format(temp.age_desc.values[0], temp.sex.values[0], \n",
    "                                              temp.tissue_desc.values[0])\n",
    "        exp_alias = 'ali-mortazavi:exp_lr_{}_{}_{}'.format(temp.age_desc.values[0], temp.sex.values[0],\n",
    "                                             temp.tissue_desc.values[0])\n",
    "\n",
    "\n",
    "    exp['assay_term_name'] = 'single-cell RNA sequencing assay'\n",
    "    exp['description'] = desc\n",
    "    exp['aliases'] = exp_alias\n",
    "    exp['lab'] = lab\n",
    "    exp['award'] = award\n",
    "    \n",
    "    cols = ['aliases', 'biosample_ontology', 'description', \n",
    "            'assay_term_name', 'lab', 'award']\n",
    "    exp = exp[cols]\n",
    "    \n",
    "    exp_sub = pd.concat([exp_sub, exp], axis=0)\n",
    "    \n",
    "    # add library\n",
    "    lib = temp.copy(deep=True)\n",
    "    \n",
    "    if long == False:\n",
    "        lib_alias = 'ali-mortazavi:library_sr_{}_{}_{}_{}'.format(temp.age_desc.values[0], temp.sex.values[0],\n",
    "                                                             temp.tissue_desc.values[0], temp.rep.values[0])\n",
    "        lib['documents'] = 'ali-mortazavi:Split-seq_computational_protocol_v1.0,ali-mortazavi:split-seq-v1'\n",
    "        lib['strand_specificity'] = 'unstranded'        \n",
    "    else: \n",
    "        lib_alias = 'ali-mortazavi:library_lr_{}_{}_{}_{}'.format(temp.age_desc.values[0], temp.sex.values[0],\n",
    "                                                     temp.tissue_desc.values[0], temp.rep.values[0])\n",
    "        lib['documents'] = 'ali-mortazavi:LR-Split-seq_computational_protocol_v1.0,ali-mortazavi:split-seq-v1,ali-mortazavi:pacbio-split-seq-v1'\n",
    "        lib['strand_specificity'] = 'forward'\n",
    "        \n",
    "    lib['aliases'] = lib_alias\n",
    "    lib['biosample'] = biosamp_alias\n",
    "    lib['nucleic_acid_term_name'] = 'RNA'\n",
    "    lib['construction_method'] = 'Parse Single Cell Whole Transcriptome Kit'\n",
    "    lib['nucleic_acid_starting_quantity'] = lib_meta.loc[lib_meta['Sample ID'] == sample_id, '# predicted nuclei'].values[0]\n",
    "    lib['nucleic_acid_starting_quantity_units'] = 'cells'\n",
    "    \n",
    "    lib['lab'] = lab\n",
    "    lib['award'] = award\n",
    "    \n",
    "    cols = ['aliases', 'biosample', 'nucleic_acid_term_name', \n",
    "            'documents', 'construction_method', 'nucleic_acid_starting_quantity',\n",
    "            'nucleic_acid_starting_quantity_units', 'strand_specificity',\n",
    "            'lab', 'award']\n",
    "    lib = lib[cols]\n",
    "    \n",
    "    lib_sub = pd.concat([lib_sub, lib], axis=0)\n",
    "    \n",
    "    # add replicate\n",
    "    rep = temp.copy(deep=True)\n",
    "    \n",
    "    if long == False:\n",
    "        rep_alias = 'ali-mortazavi:rep_sr_{}_{}_{}_{}'.format(temp.age_desc.values[0], temp.sex.values[0],\n",
    "                                                         temp.tissue_desc.values[0], temp.rep.values[0])\n",
    "    else:\n",
    "        rep_alias = 'ali-mortazavi:rep_lr_{}_{}_{}_{}'.format(temp.age_desc.values[0], temp.sex.values[0],\n",
    "                                                         temp.tissue_desc.values[0], temp.rep.values[0])\n",
    "\n",
    "    rep['aliases'] = rep_alias\n",
    "    rep['library'] = lib_alias\n",
    "    rep['experiment'] = exp_alias\n",
    "    rep['biological_replicate_number'] = temp.rep.values[0]\n",
    "    rep['technical_replicate_number'] = 1\n",
    "                                                    \n",
    "    cols = ['aliases', 'library', 'experiment', 'biological_replicate_number', 'technical_replicate_number']\n",
    "    rep = rep[cols]\n",
    "\n",
    "    rep_sub = pd.concat([rep_sub, rep], axis=0)\n",
    "    \n",
    "    # add file\n",
    "    file = temp.copy(deep=True)\n",
    "    \n",
    "    if long == False:\n",
    "        file_alias = 'ali-mortazavi:fastq_sr_{}_{}_{}_{}'.format(temp.age_desc.values[0], temp.sex.values[0],\n",
    "                                                         temp.tissue_desc.values[0], temp.rep.values[0])\n",
    "        file['read_length'] = 115\n",
    "        file['run_type'] = 'single-ended'\n",
    "        file['platform'] = 'encode:NextSeq2000'\n",
    "        \n",
    "    else:\n",
    "        file_alias = 'ali-mortazavi:fastq_lr_{}_{}_{}_{}'.format(temp.age_desc.values[0], temp.sex.values[0],\n",
    "                                                         temp.tissue_desc.values[0], temp.rep.values[0])\n",
    "        file['platform'] = 'encode:PacBio_sequel_II'\n",
    "    \n",
    "    file['aliases'] = file_alias\n",
    "    file['dataset'] = exp_alias\n",
    "    file['submitted_file_name'] = os.path.abspath(f)\n",
    "    file['replicate'] = rep_alias\n",
    "    file['file_format'] = 'fastq'\n",
    "    file['output_type'] = 'reads'\n",
    "    \n",
    "    cols = ['aliases', 'dataset', 'submitted_file_name', \n",
    "            'replicate', 'file_format', 'output_type', 'platform']\n",
    "    if long == False:\n",
    "        cols.append('read_length')\n",
    "        cols.append('run_type')\n",
    "    file = file[cols]\n",
    "    file_sub = pd.concat([file_sub, file], axis=0)\n",
    "    \n",
    "    \n",
    "# drop unnecessary columns\n",
    "drop = ['tissue_desc', 'age_desc', 'rep_desc']\n",
    "biosamp_sub.drop(drop, axis=1, inplace=True)\n",
    "\n",
    "# for exp, also drop duplicate aliases \n",
    "exp_sub.drop_duplicates(inplace=True)\n",
    "\n",
    "# save each table\n",
    "if not long:\n",
    "    opref = '{}_sr'.format(opref)\n",
    "else:\n",
    "    opref = '{}_lr'.format(opref)    \n",
    "    \n",
    "# biosample\n",
    "fname = opref+'_biosample.tsv'\n",
    "biosample.to_csv(fname, index=False, sep='\\t')\n",
    "\n",
    "# experiment\n",
    "fname = opref+'_experiment.tsv'\n",
    "exp.to_csv(fname, index=False, sep='\\t')\n",
    "\n",
    "# library\n",
    "fname = opref+'_library.tsv'\n",
    "lib.to_csv(fname, index=False, sep='\\t')\n",
    "\n",
    "# replicate\n",
    "fname = opref+'_rep.tsv'\n",
    "rep.to_csv(fname, index=False, sep='\\t')\n",
    "\n",
    "# file\n",
    "fname = opref+'_file.tsv'\n",
    "file.to_csv(fname, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "velvet-worthy",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_ontology</th>\n",
       "      <th>organism</th>\n",
       "      <th>subcellular_fraction_term_name</th>\n",
       "      <th>donor</th>\n",
       "      <th>source</th>\n",
       "      <th>sex</th>\n",
       "      <th>model_organism_age</th>\n",
       "      <th>model_organism_age_units</th>\n",
       "      <th>rep</th>\n",
       "      <th>date_obtained</th>\n",
       "      <th>description</th>\n",
       "      <th>aliases</th>\n",
       "      <th>lab</th>\n",
       "      <th>award</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>/biosample-types/tissue_UBERON_0002369/</td>\n",
       "      <td>/organisms/mouse/</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>encode:Castaneus</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>male</td>\n",
       "      <td>18-20</td>\n",
       "      <td>month</td>\n",
       "      <td>2</td>\n",
       "      <td>7/8/21</td>\n",
       "      <td>B6Cast F1 P18-20mo male adrenal 8</td>\n",
       "      <td>ali-mortazavi:biosample_P18-20mo_male_adrenal_2</td>\n",
       "      <td>ali-mortazavi</td>\n",
       "      <td>UM1HG009443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>/biosample-types/tissue_UBERON_0002369/</td>\n",
       "      <td>/organisms/mouse/</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>encode:Castaneus</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>female</td>\n",
       "      <td>10</td>\n",
       "      <td>day</td>\n",
       "      <td>1</td>\n",
       "      <td>7/8/21</td>\n",
       "      <td>B6Cast F1 P10 female adrenal 7</td>\n",
       "      <td>ali-mortazavi:biosample_P10_female_adrenal_1</td>\n",
       "      <td>ali-mortazavi</td>\n",
       "      <td>UM1HG009443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>/biosample-types/tissue_UBERON_0002369/</td>\n",
       "      <td>/organisms/mouse/</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>encode:Castaneus</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>female</td>\n",
       "      <td>10</td>\n",
       "      <td>day</td>\n",
       "      <td>2</td>\n",
       "      <td>7/8/21</td>\n",
       "      <td>B6Cast F1 P10 female adrenal 8</td>\n",
       "      <td>ali-mortazavi:biosample_P10_female_adrenal_2</td>\n",
       "      <td>ali-mortazavi</td>\n",
       "      <td>UM1HG009443</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        biosample_ontology           organism  \\\n",
       "0  /biosample-types/tissue_UBERON_0002369/  /organisms/mouse/   \n",
       "0  /biosample-types/tissue_UBERON_0002369/  /organisms/mouse/   \n",
       "0  /biosample-types/tissue_UBERON_0002369/  /organisms/mouse/   \n",
       "\n",
       "  subcellular_fraction_term_name             donor                  source  \\\n",
       "0                        nucleus  encode:Castaneus  /sources/jackson-labs/   \n",
       "0                        nucleus  encode:Castaneus  /sources/jackson-labs/   \n",
       "0                        nucleus  encode:Castaneus  /sources/jackson-labs/   \n",
       "\n",
       "      sex model_organism_age model_organism_age_units rep date_obtained  \\\n",
       "0    male              18-20                    month   2        7/8/21   \n",
       "0  female                 10                      day   1        7/8/21   \n",
       "0  female                 10                      day   2        7/8/21   \n",
       "\n",
       "                         description  \\\n",
       "0  B6Cast F1 P18-20mo male adrenal 8   \n",
       "0     B6Cast F1 P10 female adrenal 7   \n",
       "0     B6Cast F1 P10 female adrenal 8   \n",
       "\n",
       "                                           aliases            lab        award  \n",
       "0  ali-mortazavi:biosample_P18-20mo_male_adrenal_2  ali-mortazavi  UM1HG009443  \n",
       "0     ali-mortazavi:biosample_P10_female_adrenal_1  ali-mortazavi  UM1HG009443  \n",
       "0     ali-mortazavi:biosample_P10_female_adrenal_2  ali-mortazavi  UM1HG009443  "
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biosamp_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "sunset-joining",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aliases</th>\n",
       "      <th>biosample_ontology</th>\n",
       "      <th>description</th>\n",
       "      <th>assay_term_name</th>\n",
       "      <th>lab</th>\n",
       "      <th>award</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ali-mortazavi:exp_sr_P18-20mo_male_adrenal</td>\n",
       "      <td>/biosample-types/tissue_UBERON_0002369/</td>\n",
       "      <td>Short-read Split-seq B6Cast F1 P18-20mo male a...</td>\n",
       "      <td>single-cell RNA sequencing assay</td>\n",
       "      <td>ali-mortazavi</td>\n",
       "      <td>UM1HG009443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ali-mortazavi:exp_sr_P10_female_adrenal</td>\n",
       "      <td>/biosample-types/tissue_UBERON_0002369/</td>\n",
       "      <td>Short-read Split-seq B6Cast F1 P10 female adrenal</td>\n",
       "      <td>single-cell RNA sequencing assay</td>\n",
       "      <td>ali-mortazavi</td>\n",
       "      <td>UM1HG009443</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      aliases  \\\n",
       "0  ali-mortazavi:exp_sr_P18-20mo_male_adrenal   \n",
       "0     ali-mortazavi:exp_sr_P10_female_adrenal   \n",
       "\n",
       "                        biosample_ontology  \\\n",
       "0  /biosample-types/tissue_UBERON_0002369/   \n",
       "0  /biosample-types/tissue_UBERON_0002369/   \n",
       "\n",
       "                                         description  \\\n",
       "0  Short-read Split-seq B6Cast F1 P18-20mo male a...   \n",
       "0  Short-read Split-seq B6Cast F1 P10 female adrenal   \n",
       "\n",
       "                    assay_term_name            lab        award  \n",
       "0  single-cell RNA sequencing assay  ali-mortazavi  UM1HG009443  \n",
       "0  single-cell RNA sequencing assay  ali-mortazavi  UM1HG009443  "
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "essential-field",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aliases</th>\n",
       "      <th>biosample</th>\n",
       "      <th>nucleic_acid_term_name</th>\n",
       "      <th>documents</th>\n",
       "      <th>construction_method</th>\n",
       "      <th>nucleic_acid_starting_quantity</th>\n",
       "      <th>nucleic_acid_starting_quantity_units</th>\n",
       "      <th>strand_specificity</th>\n",
       "      <th>lab</th>\n",
       "      <th>award</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ali-mortazavi:library_sr_P18-20mo_male_adrenal_2</td>\n",
       "      <td>ali-mortazavi:biosample_P18-20mo_male_adrenal_2</td>\n",
       "      <td>RNA</td>\n",
       "      <td>ali-mortazavi:Split-seq_computational_protocol...</td>\n",
       "      <td>Parse Single Cell Whole Transcriptome Kit</td>\n",
       "      <td>3125</td>\n",
       "      <td>cells</td>\n",
       "      <td>unstranded</td>\n",
       "      <td>ali-mortazavi</td>\n",
       "      <td>UM1HG009443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ali-mortazavi:library_sr_P10_female_adrenal_1</td>\n",
       "      <td>ali-mortazavi:biosample_P10_female_adrenal_1</td>\n",
       "      <td>RNA</td>\n",
       "      <td>ali-mortazavi:Split-seq_computational_protocol...</td>\n",
       "      <td>Parse Single Cell Whole Transcriptome Kit</td>\n",
       "      <td>1562</td>\n",
       "      <td>cells</td>\n",
       "      <td>unstranded</td>\n",
       "      <td>ali-mortazavi</td>\n",
       "      <td>UM1HG009443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ali-mortazavi:library_sr_P10_female_adrenal_2</td>\n",
       "      <td>ali-mortazavi:biosample_P10_female_adrenal_2</td>\n",
       "      <td>RNA</td>\n",
       "      <td>ali-mortazavi:Split-seq_computational_protocol...</td>\n",
       "      <td>Parse Single Cell Whole Transcriptome Kit</td>\n",
       "      <td>1562</td>\n",
       "      <td>cells</td>\n",
       "      <td>unstranded</td>\n",
       "      <td>ali-mortazavi</td>\n",
       "      <td>UM1HG009443</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            aliases  \\\n",
       "0  ali-mortazavi:library_sr_P18-20mo_male_adrenal_2   \n",
       "0     ali-mortazavi:library_sr_P10_female_adrenal_1   \n",
       "0     ali-mortazavi:library_sr_P10_female_adrenal_2   \n",
       "\n",
       "                                         biosample nucleic_acid_term_name  \\\n",
       "0  ali-mortazavi:biosample_P18-20mo_male_adrenal_2                    RNA   \n",
       "0     ali-mortazavi:biosample_P10_female_adrenal_1                    RNA   \n",
       "0     ali-mortazavi:biosample_P10_female_adrenal_2                    RNA   \n",
       "\n",
       "                                           documents  \\\n",
       "0  ali-mortazavi:Split-seq_computational_protocol...   \n",
       "0  ali-mortazavi:Split-seq_computational_protocol...   \n",
       "0  ali-mortazavi:Split-seq_computational_protocol...   \n",
       "\n",
       "                         construction_method  nucleic_acid_starting_quantity  \\\n",
       "0  Parse Single Cell Whole Transcriptome Kit                            3125   \n",
       "0  Parse Single Cell Whole Transcriptome Kit                            1562   \n",
       "0  Parse Single Cell Whole Transcriptome Kit                            1562   \n",
       "\n",
       "  nucleic_acid_starting_quantity_units strand_specificity            lab  \\\n",
       "0                                cells         unstranded  ali-mortazavi   \n",
       "0                                cells         unstranded  ali-mortazavi   \n",
       "0                                cells         unstranded  ali-mortazavi   \n",
       "\n",
       "         award  \n",
       "0  UM1HG009443  \n",
       "0  UM1HG009443  \n",
       "0  UM1HG009443  "
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lib_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "sixth-cradle",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aliases</th>\n",
       "      <th>library</th>\n",
       "      <th>experiment</th>\n",
       "      <th>biological_replicate_number</th>\n",
       "      <th>technical_replicate_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ali-mortazavi:rep_sr_P18-20mo_male_adrenal_2</td>\n",
       "      <td>ali-mortazavi:library_sr_P18-20mo_male_adrenal_2</td>\n",
       "      <td>ali-mortazavi:exp_sr_P18-20mo_male_adrenal</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ali-mortazavi:rep_sr_P10_female_adrenal_1</td>\n",
       "      <td>ali-mortazavi:library_sr_P10_female_adrenal_1</td>\n",
       "      <td>ali-mortazavi:exp_sr_P10_female_adrenal</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ali-mortazavi:rep_sr_P10_female_adrenal_2</td>\n",
       "      <td>ali-mortazavi:library_sr_P10_female_adrenal_2</td>\n",
       "      <td>ali-mortazavi:exp_sr_P10_female_adrenal</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        aliases  \\\n",
       "0  ali-mortazavi:rep_sr_P18-20mo_male_adrenal_2   \n",
       "0     ali-mortazavi:rep_sr_P10_female_adrenal_1   \n",
       "0     ali-mortazavi:rep_sr_P10_female_adrenal_2   \n",
       "\n",
       "                                            library  \\\n",
       "0  ali-mortazavi:library_sr_P18-20mo_male_adrenal_2   \n",
       "0     ali-mortazavi:library_sr_P10_female_adrenal_1   \n",
       "0     ali-mortazavi:library_sr_P10_female_adrenal_2   \n",
       "\n",
       "                                   experiment  biological_replicate_number  \\\n",
       "0  ali-mortazavi:exp_sr_P18-20mo_male_adrenal                            2   \n",
       "0     ali-mortazavi:exp_sr_P10_female_adrenal                            1   \n",
       "0     ali-mortazavi:exp_sr_P10_female_adrenal                            2   \n",
       "\n",
       "   technical_replicate_number  \n",
       "0                           1  \n",
       "0                           1  \n",
       "0                           1  "
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rep_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "eight-liabilities",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aliases</th>\n",
       "      <th>dataset</th>\n",
       "      <th>submitted_file_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>file_format</th>\n",
       "      <th>output_type</th>\n",
       "      <th>platform</th>\n",
       "      <th>read_length</th>\n",
       "      <th>run_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ali-mortazavi:fastq_sr_P18-20mo_male_adrenal_2</td>\n",
       "      <td>ali-mortazavi:exp_sr_P18-20mo_male_adrenal</td>\n",
       "      <td>/Users/fairliereese/Documents/programming/mort...</td>\n",
       "      <td>ali-mortazavi:rep_sr_P18-20mo_male_adrenal_2</td>\n",
       "      <td>fastq</td>\n",
       "      <td>reads</td>\n",
       "      <td>encode:NextSeq2000</td>\n",
       "      <td>115</td>\n",
       "      <td>single-ended</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ali-mortazavi:fastq_sr_P10_female_adrenal_1</td>\n",
       "      <td>ali-mortazavi:exp_sr_P10_female_adrenal</td>\n",
       "      <td>/Users/fairliereese/Documents/programming/mort...</td>\n",
       "      <td>ali-mortazavi:rep_sr_P10_female_adrenal_1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>reads</td>\n",
       "      <td>encode:NextSeq2000</td>\n",
       "      <td>115</td>\n",
       "      <td>single-ended</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ali-mortazavi:fastq_sr_P10_female_adrenal_2</td>\n",
       "      <td>ali-mortazavi:exp_sr_P10_female_adrenal</td>\n",
       "      <td>/Users/fairliereese/Documents/programming/mort...</td>\n",
       "      <td>ali-mortazavi:rep_sr_P10_female_adrenal_2</td>\n",
       "      <td>fastq</td>\n",
       "      <td>reads</td>\n",
       "      <td>encode:NextSeq2000</td>\n",
       "      <td>115</td>\n",
       "      <td>single-ended</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          aliases  \\\n",
       "0  ali-mortazavi:fastq_sr_P18-20mo_male_adrenal_2   \n",
       "0     ali-mortazavi:fastq_sr_P10_female_adrenal_1   \n",
       "0     ali-mortazavi:fastq_sr_P10_female_adrenal_2   \n",
       "\n",
       "                                      dataset  \\\n",
       "0  ali-mortazavi:exp_sr_P18-20mo_male_adrenal   \n",
       "0     ali-mortazavi:exp_sr_P10_female_adrenal   \n",
       "0     ali-mortazavi:exp_sr_P10_female_adrenal   \n",
       "\n",
       "                                 submitted_file_name  \\\n",
       "0  /Users/fairliereese/Documents/programming/mort...   \n",
       "0  /Users/fairliereese/Documents/programming/mort...   \n",
       "0  /Users/fairliereese/Documents/programming/mort...   \n",
       "\n",
       "                                      replicate file_format output_type  \\\n",
       "0  ali-mortazavi:rep_sr_P18-20mo_male_adrenal_2       fastq       reads   \n",
       "0     ali-mortazavi:rep_sr_P10_female_adrenal_1       fastq       reads   \n",
       "0     ali-mortazavi:rep_sr_P10_female_adrenal_2       fastq       reads   \n",
       "\n",
       "             platform  read_length      run_type  \n",
       "0  encode:NextSeq2000          115  single-ended  \n",
       "0  encode:NextSeq2000          115  single-ended  \n",
       "0  encode:NextSeq2000          115  single-ended  "
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "national-encoding",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_metadata(f):\n",
    "    tissue, age, sex, rep = f.split('_')\n",
    "    rep = rep.split('.')[0]\n",
    "    tissue = tissue.split('/')[-1]\n",
    "    return tissue, age, sex, rep\n",
    "\n",
    "def get_sample_id(f):\n",
    "    s = f.split('/')[-1]\n",
    "    s = s.split('.')[0]\n",
    "    return s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "appointed-showcase",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "dense-cleaners",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/fairliereese/miniconda3/lib/python3.7/site-packages/ipykernel_launcher.py'"
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sys.argv[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "spread-consequence",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
